### ISSUE TITLE
Proposal: Record which user created an object

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#13306]
Currently no creator information is recorded at the time an object is created. The creator may be:
- an end user calling the api (authenticated via the Authenticator interfaces)
- a controller calling the api (authenticated via Authenticator)
- a core controller locally invoking the storage interfaces (bypasses authentication)
- someone directly editing etcd (repair process or object created in the past).

It is a common requirement to record audit information with objects that allow UIs and administrators to determine the actor involved in a transaction. The information may be recorded in a side channel (an audit log), but end users may also wish to understand who created a particular object.

Use cases:
- Allow clients to determine which user made a change
- Allow an administrator to reconstruct an audit record for who created which resources were created in the system

In the first two scenarios above, the user is clearly identified (`GetUserName()` on the user resolved by the authenticator, which rest.Storage has access to).  In the third, this is effectively what we call a "system" user, and could be assigned an arbitrary username that does not collide with real user names, or left blank.  In the last case there is probably a user name, but no mechanism whereby it is provided.
#### Inferential attribution
- Actions may not be undertaken by a user, but on behalf of a user
  - the replication controller process creates pods on behalf of the user who created a replication controller, but if another user edits the replication controller, both users may have a causal relationship to the pods created
  - a virtual resource may run as elevated privileges against another api and allow a user to create a resource they normally don't have permission to create - both the end user and the api process have a causal relationship to the underlying resource
  - a higher privilege user may request to run as another user via **impersonation** - the actor is the higher privilege user, but they are explicitly wishing to be recorded as acting as the impersonated user
- Actions may be separated in time and space from their cause
  - a user may delete a node, which results in a pod being deleted, which results in a pod being created by a replication controller

Recommendation: we should scope the attribution of creation to the caller (direct creator), and ensure that an external "audit log" could be constructed that can infer chain of attribution for complex cases.
#### Creating child resources

In general, it may be desirable for high level constructs (deployments, scheduled jobs) that create lower level resources (replication controllers, jobs) to propagate creation attribution. However, that raises a security risk that the creating component can lie about the source of a creation. Since an audit trail can be reconstructed between parent and subordinate resources by observing API changes, it is not necessary to propagate that info.
#### Referencing a user

We currently define a `user name` as a string that represents a unique user. That value is expected to be stable over the time periods the admin desires (only admin action would result in a user name changing). Under the covers, the user name may refer to many different types of users - the [authentication design](https://github.com/kubernetes/kubernetes/blob/master/docs/design/access.md) identifies two:
- User
  - ServiceAccount - a namespaced resource that represents a unique automated identity
    - In the future, this account may be backed by an external user
  - ExternalUser - via an external system
    - May be a human or an external service account

In addition, we have an implicit type:
- SystemUser - created implicitly in the infrastructure to represent distinct process identity prior to bootstrapping of service accounts or external users
  - Which user represents the code in the master that creates the default service account?

The `Authenticator` interface requires all of the above users to be mapped into a single user name string before filtering through the system. This string is the `userName` and must be unique across all sources of authentication.

It is assumed that the audit system and other client systems understand the mapping between the various authentication sources and `userName`.  These mappings may differ between different installations.  It should be possible for a client or third party to provide a `userName` and retrieve an object (or objects?) representing the source of the user name. 

As an example, Kube must overlay the service account user identifier onto the external authenticator source. Today, that is enforced via a unique prefix `system:serviceaccount:` that is reserved. In the future, we may map `ServiceAccount` to an external user, in which case the unique prefix may not be needed. OpenShift, which needs to manage users internally as well as externally, uses a further system of subdivision of namespace to assign a `userName` via indirection.
### Proposed change

When objects are created through rest.Storage, record the identifying user info onto the object. Add `creationUserName string` to `ObjectMeta` and assign it the `GetUserName()` value provided by the Authenticator. This value may not be changed, but may be unset if authentication is disabled.

Implications for clients:
- Clients will want to show "pretty" names to end users (`userName` may be email, but may also be a unique identifier or LDAP DN or short name). Clients that wish to do that will need a way to map `userName` to a useful value efficiently and in bulk (it's assumed that the "pretty name" for a user changes at very infrequent intervals)
- Clients may be tempted to assume the `userName` can be parsed. This limits the admin's ability to change this value. Instruct clients not to assume the `userName` value will be human readable or that it follows a parseable pattern.

Other implications:
- Clients should not be allowed to set the `creationUserName`. 
- Administrators may wish to migrate old data to assign a `creationUserName`.

**Comment by smarterclayton** [comment]
@liggitt @erictune @deads2k @jwforres @kubernetes/kube-api  @kubernetes/kube-iam

**Comment by jwforres** [comment]
@smarterclayton are you proposing that the API would not return the pretty name along with the username, but instead the clients would have to make secondary calls to get the pretty names for the users?

**Comment by liggitt** [comment]
there are no pretty names inherent in the user.Info interface, which is what would be used to populate this

**Comment by smarterclayton** [comment]
And I punted on making the name an object ref because we don't even have a
formal object kind hierarchy for users (in Kube) and it's hard to impose
just one.

On Aug 28, 2015, at 3:40 PM, Jordan Liggitt <notifications@github.com>
wrote:

there are no pretty names inherent in the user.Info interface, which is
what would be used to populate this

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13306#issuecomment-135869344>
.

**Comment by deads2k** [comment]
Is knowing a creator valuable without knowing the updaters?  I understand that creator is easy to track since there's guaranteed to only be one of them, but not knowing whether it's changed is kind of a big deal.  What about a `lastUpdater` marker to indicate that something outside of `status` has changed?

**Comment by smarterclayton** [comment]
Updater, for the purposes of this discussion, is a more complex creator.
So regardless of whether updater is added, the details about creator here
establish patterns that also apply to updater.

Tracking last update user depends on sorting out last update time as well
(which is extremely difficult to do correctly).  Also, updater on the
object is less useful without the ability to walk history which we are
waiting for etcd3 for.

On Aug 31, 2015, at 8:21 AM, David Eads <notifications@github.com> wrote:

Is knowing a creator valuable without knowing the updaters? I understand
that creator is easy to track since there's guaranteed to only be one of
them, but not knowing whether it's changed is kind of a big deal. What
about a lastUpdater marker to indicate that something outside of status has
changed?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13306#issuecomment-136356721>
.

**Comment by smarterclayton** [comment]
Added a proposed implementation in #13925

**Comment by erictune** [comment]
I'm also worried that Is knowing a creator valuable without knowing the updaters?  Even knowing lastUpdater is not enough.  Any updater could have made the change that the auditor wishes to detect.

@smarterclayton I'm not sure how to reconcile your comment that the audit log can be reconstructed with your comment that finding the last updater is hard because time is hard.  Doesn't that imply that reconstructing even the indirect creator is hard because time is hard?  What if there is a change to an RC template at about the same time a pod was created.  Hard to be sure which template the manager used.

Breaking down by resources:
- pods: we recommend that these be created indirectly, so these will rarely have a human creationUserName on them.
- rcs: these should be created by deployments most of the time, IIUC.
- deployments: these are supposed to be long lived and get updated many times, perhaps by different humans.  So, while the creation user is valid, we don't know updater.
- services: it seems like just knowing who created a service is not too useful.  You want to know if a certain pod in question was ever expose via a particular services.  For which you need a different kind of logs?
- secrets: for this you probably want record of reads more so than a record of creations?
- jobs: creation user is probably useful for these.

**Comment by smarterclayton** [comment]
Audit is one facet of this - knowing that Bob created service DB is often
important if I'm Alice looking at a namespace - she knows who to ask about
"db".

As far as inference, I agree that more data would be important, and I
wasn't proposing creationUserName solves all of the problems.  However, it
is the start of the solution and also provides valuable interaction data
with the rest of the system.  I think we *should* figure out how to tie
updater to resources, as well as last modification time, as well as add
detailed audit and inferential tools around the platform to make those
reliable.  I felt that creationUserName stood on its own in that virtually
all systems where multiple people interact, the record of the creator and
modifier is useful for multiple purposes.  I just did not want to
explicitly block creation on modification, because creation at least can be
tied to an entity today (even if that entity is an "on behalf of"
relationship).

A pod that is created by a user is really important (more so than a pod
created by an rc, more so than an rc created by a user is more important
than an rc created by a deployment).


On Fri, Sep 18, 2015 at 3:31 PM, Eric Tune <notifications@github.com> wrote:

> I'm also worried that Is knowing a creator valuable without knowing the
> updaters? Even knowing lastUpdater is not enough. Any updater could have
> made the change that the auditor wishes to detect.
>
> @smarterclayton <https://github.com/smarterclayton> I'm not sure how to
> reconcile your comment that the audit log can be reconstructed with your
> comment that finding the last updater is hard because time is hard. Doesn't
> that imply that reconstructing even the indirect creator is hard because
> time is hard? What if there is a change to an RC template at about the same
> time a pod was created. Hard to be sure which template the manager used.
>
> Breaking down by resources:
>
>    - pods: we recommend that these be created indirectly, so these will
>    rarely have a human creationUserName on them.
>    - rcs: these should be created by deployments most of the time, IIUC.
>    - deployments: these are supposed to be long lived and get updated
>    many times, perhaps by different humans. So, while the creation user is
>    valid, we don't know updater.
>    - services: it seems like just knowing who created a service is not
>    too useful. You want to know if a certain pod in question was ever expose
>    via a particular services. For which you need a different kind of logs?
>    - secrets: for this you probably want record of reads more so than a
>    record of creations?
>    - jobs: creation user is probably useful for these.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/13306#issuecomment-141545477>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
More comments on this?

**Comment by lavalamp** [comment]
It seems to me that audit logging and storing the creator/updator of an object are two different use cases and it might make sense to start out by just doing the latter.

**Comment by smarterclayton** [comment]
In general I agree - do you see that as being in schema (part of
objectmeta) or out (annotation controlled by admission control)?

On Oct 8, 2015, at 7:03 PM, Daniel Smith <notifications@github.com> wrote:

It seems to me that audit logging and storing the creator/updator of an
object are two different use cases and it might make sense to start out by
just doing the latter.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13306#issuecomment-146711651>
.

**Comment by smarterclayton** [comment]
@erictune @lavalamp I have one for, one no-response.  I'd like to push forward on this and get closure so I can plan for it our audit control stories.  Strong disagreement is welcome, as long as it's prompt.

**Comment by lavalamp** [comment]
@smarterclayton Sorry I missed your response.

I think eventually this clearly has to go in object meta. But object meta is hard to change later, and I don't know that we know how exactly we want to refer to users. I think in this case having an annotation first is probably the way to go. When we know it's right, we can move it to object meta. That has its own pitfalls, though.

**Comment by smarterclayton** [comment]
Ok - for OpenShift we'll probably try an annotation admission controller that enforces this value on everything (add, remove, update) and prevents edits.  Once I have something working I'll open a PR to kube.

**Comment by erictune** [comment]
Sorry for late response.  I don't think we should store creator in metadata. 
I commented on #13925 just now.

**Comment by deads2k** [comment]
> Sorry for late response. I don't think we should store creator in metadata. 
> I commented on #13925 just now.

@smarterclayton This came up at sig-auth.  We couldn't think of a reasonable way to use the information in this field safely.  For any mutable resource, as soon as its dirtied the information isn't valuable anymore.  Pretty much every resource gets dirtied by a controller.  The temptation is create two classes of users, but that seems like we're trying to shoehorn it in where it doesn't fit.

Do we have a specific need for this? The persistent volume binding controller came up recently, but that is very focused (they could wire their own resource) and likely a mis-use anyway.

@kubernetes/sig-auth

**Comment by rikatz** [comment]
Isn't this already covered by enabling the audit log on Kubernetes?

An example is when a user creates an object on a namespace, we get something like this:

2016-12-09T13:51:28.161672089Z AUDIT: id="5722fecc-7493-4688-a888-a5313bf019e6" ip="10.34.144.60" method="PUT" user="user1" as="<self>" namespace="gscvi-t" uri="/apis/extensions/v1beta1/namespaces/gscvi-t/deployments/appfla"

Isn't this enough for this issue?

I would like to see a more verbose in auditing, like when an object is created / patched / changed / deleted, auditing the old and the new value.

I was working on some kind of auditing like this directly on ETCD (watching key changes) but I don't think this is enough. 

An example, for my needs: When a user changes a network policy, I do have to audit this, as this could turn an app vulnerable to unwanted access.

**Comment by deads2k** [comment]
> rikatz commented 2 hours ago
> Isn't this already covered by enabling the audit log on Kubernetes?
> 
> An example is when a user creates an object on a namespace, we get something like this:
> 
> 2016-12-09T13:51:28.161672089Z AUDIT: id="5722fecc-7493-4688-a888-a5313bf019e6" ip="10.34.144.60" method="PUT" user="user1" as="" namespace="gscvi-t" uri="/apis/extensions/v1beta1/namespaces/gscvi-t/deployments/appfla"
> 
> Isn't this enough for this issue?

No.  This is about tracking it inside the object, so that future API callers have access to the information via the API.

I still don't think this is a good idea.  It's too open to abuse and unexpected results.

**Comment by nikhiljindal** [comment]
This came up while discussing https://github.com/kubernetes/kubernetes/issues/35254 for federation.

> from @smarterclayton 
> Ok - for OpenShift we'll probably try an annotation admission controller that enforces this value on everything (add, remove, update) and prevents edits. Once I have something working I'll open a PR to kube.
> 
@smarterclayton are you using this in OpenShift?

Lot of the discussion here is split across this issue and the corresponding PR (https://github.com/kubernetes/kubernetes/pull/13925).

The main thing this seems to be blocked on is to collect more use cases where this will be useful. I see the main motivation to start this proposal was to use it for audit logging. https://github.com/kubernetes/kubernetes/pull/27087 has now improved that. 
Another [use case](https://github.com/kubernetes/kubernetes/pull/13925#issuecomment-175136088) that came up was to use the resource owner for scheduling decisions.

Some other discussion points on which we seem to have agreed are:
* How to enforce that users can not update this field: Add an admission controller.
* Do we need a field for user who last updated the resource as well: Yes, but should start with `creationUserName`.
* Is just the name enough or we need UID as well: Start with name. Can add uid later.

**Comment by deads2k** [comment]
> @smarterclayton are you using this in OpenShift?

No, we are not.

**Comment by timothysc** [comment]
Another use case is idempotent installation of manifests.  xref: https://github.com/kubernetes/kubernetes/pull/47081

**Comment by smarterclayton** [comment]
we talked about the historical use case of having something scraping this
and recording the side history such that you could look, but that would
require us to record the actual mutating user into etcd as well.  So either
way, we need to pass some small chunk of info down in the final object.

On Wed, Jun 7, 2017 at 10:40 AM, Timothy St. Clair <notifications@github.com

**Comment by liggitt** [comment]
auditing provides this information for logging purposes. there are enough issues with PII and potential for misuse that I don't see a reasonable way to do this generally for all objects
